src/core/algorithms/ind/faida/faida.cpp

```diff
@@ -0,0 +1,232 @@
+    RegisterOption(Option{&hll_accuracy_, kHllAccuracy, kDHllAccuracy, 0.001});
+    RegisterOption(Option{&detect_nary_, kFindNary, kDFindNary, true});
+    RegisterOption(config::ThreadNumberOpt(&number_of_threads_));
+
+    MakeOptionsAvailable({kSampleSize});
+}
+
+void Faida::MakeExecuteOptsAvailable() {
+    using namespace config::names;
+    MakeOptionsAvailable({kFindNary, kHllAccuracy, config::ThreadNumberOpt.GetName()});
+}
+
+void Faida::LoadDataInternal() {
+    auto start_time = std::chrono::system_clock::now();
+
+    data_ = faida::Preprocessor::CreateHashedStores(input_table_set_.name, input_table_set_.parsers,
+                                                    sample_size_);
+
+    for (auto const& store : data_->GetStores()) {
+        auto const& schema = store->GetSchema();
+
+        DatasetInfo ds_info;
+        ds_info.table_name = schema->GetName();
+        for (auto const& column : schema->GetColumns()) {
+            ds_info.header.push_back(column->GetName());
+        }
+
+        dataset_order_.emplace_back(std::move(ds_info));
+    }
+
+    auto const prep_milliseconds = std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::system_clock::now() - start_time);
+    prepr_time_ = prep_milliseconds.count();
+    LOG(INFO) << "Preprocessing time: " << prepr_time_;
+}
+
+void Faida::ResetState() {
+    result_.clear();
+    inclusion_tester_.reset();
+    insert_time_ = 0;
+    check_time_ = 0;
+}
+
+std::vector<std::shared_ptr<faida::SimpleCC>> Faida::CreateUnaryCCs(
+        faida::Preprocessor const& data) const {
+    std::vector<std::shared_ptr<SimpleCC>> combinations;
+
+    int const index = 0;
+    for (unsigned table_num = 0; table_num < data.GetStores().size(); table_num++) {
+        auto const& store = data.GetStores()[table_num];
+        size_t const num_columns = store->GetSchema()->GetNumColumns();
+
+        for (unsigned col_idx = 0; col_idx < num_columns; col_idx++) {
+            if (store->IsConstantCol(col_idx) || store->IsNullCol(col_idx)) {
+                LOG(INFO) << "Got null or constant col with idx " << col_idx;
+                continue;
+            }
+            std::vector<int> col_indices(1, col_idx);
+            combinations.emplace_back(
+                    std::make_shared<SimpleCC>(table_num, std::move(col_indices), index));
+        }
+    }
+
+    return combinations;
+}
+
+std::vector<faida::SimpleIND> Faida::CreateUnaryINDCandidates(
+        std::vector<std::shared_ptr<SimpleCC>> const& combinations) const {
+    std::vector<SimpleIND> candidates;
+    candidates.reserve(combinations.size() * combinations.size());
+
+    for (auto left_it = combinations.begin(); left_it != combinations.end(); left_it++) {
+        for (auto right_it = combinations.begin(); right_it != combinations.end(); right_it++) {
+            if (left_it != right_it) {
+                candidates.emplace_back(*left_it, *right_it);
+            }
+        }
+    }
```

Comment:
Thanks for the good point, comparison by pointer is really something strange. But also, since the `SimpleIND` class has a comparison operator defined, we can use this operator. It also seems like this is exactly what we need semantically.
It will look like this:
```suggestion
    for (auto const& left : combinations) {
        for (auto const& right : combinations) {
            if (left != right) {
                candidates.emplace_back(left, right);
            }
        }
    }
```

Reply:
`left` and `right` here are  instances of `SimpleCC` which is essentially a vector of column indices, and the comparison operator compares this vectors per-element, which is not so fast. The size of `combinations` here equals to the total count of columns in all tables, so the comparison operation is perfrormed square of number of columns times.